### PR TITLE
Added ptrace specialisation tests and fixed a bug for ptrace_dense.

### PR DESCRIPTION
**Pull request by AGaliciaMartinez**
**Checklist**
- [x] Contributions to qutip should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/).

**Description**
Added ptrace specialisation tests. While adding these tests a bug came out with `ptrace_dense` which is now fixed. The problem was it did not work for `sel=[]`, that is, no selection and hence the full trace was required. The bug comes from how `numpy` works as the output of np.prod([]) is float instead of integer. 

I wonder though what should be the behaviour for `qobj.ptrace([])`. Should it not be the same as `qobj.trace()`?  `qobj.ptrace([])` currently returns a `Qobj` of shape 1x1 whereas `qobj.trace()` returns a complex number.

**Changelog**
Fixed error with ptrace when the input and output are dense and `sel=[]`.
Added specialisations tests for ptrace.
Improved `ptrace` tests to test for edge cases (trace all and trace none).
Added a few more checks to catch invalid dims and raise ValueError.

### CODE REVIEW

File: qutip/tests/core/test_ptrace.py [hunk 1]
```diff
@@ -55,7 +55,7 @@ def expected(qobj, sel):
             continue
         tmp_dims = (before, dim, after) * 2
         out = np.einsum('aibcid->abcd', out.reshape(tmp_dims))
-    return qutip.Qobj(out.reshape(new_shape), dims=dims)
+    return qutip.Qobj(out.reshape(new_shape), dims=dims, type='oper')
```
Comment (AGaliciaMartinez) [inline qutip/tests/core/test_ptrace.py:58]: It seems that Qobj is not able to infer the type from an empty dims (dims=[[][]]).

File: qutip/tests/core/data/test_ptrace.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    def generate_incorrect_dims_raises(self, metafunc):
+        self.generate_mathematically_correct(metafunc)
+
+    @pytest.mark.parametrize('sel',
+                             [[2, 10], [-1, 2]],
+                             ids=['sel_value_larger_than_dims',
+                                  'sel_value_negative',
+                                 ])
+    def test_incorrect_sel_raises(self, op, data_m, out_type, sel):
+        with pytest.raises(IndexError):
```
Comment (AGaliciaMartinez) [inline qutip/tests/core/data/test_ptrace.py:92]: Should we raise ValueError instead?

Comment (hodgestar) [reply qutip/tests/core/data/test_ptrace.py:92]: Is `sel` not conceptually used as `dims[sel]` effectively? If so, that looks like an incorrect index?

File: qutip/tests/core/test_ptrace.py (hunk 1 above)
Comment (hodgestar) [reply qutip/tests/core/test_ptrace.py:58]: That seems not completely unexpected.

File: qutip/tests/core/test_ptrace.py [hunk 3]
```diff
@@ -120,29 +120,42 @@ def test_ptrace_fails_on_invalid_input(state, selection, exception):
...
+
+@pytest.mark.parametrize('sel', [[], [0, 1, 2], [0], [1], [1, 0], [0, 2]],
+                        ids=['trace_all',
+                             'trace_none',
+                             'trace_one',
+                             'trace_one_2',
+                             'trace_multiple',
+                             'trace_multiple_not_sorted',
+                            ])
+def test_ptrace_dm(dtype, sel):
```
Comment (hodgestar) [inline qutip/tests/core/test_ptrace.py:145]:
```suggestion
def test_ptrace_rand_dm(dtype, sel):
```
Just to match the name of `_rand_ket` above.

**Review by hodgestar — CHANGES_REQUESTED** [review]
Left some small suggestions but woot for adding tests and finding the differing behaviour.

Regarding `qobj.ptrace([])` vs `qobj.trace()`: I *think* the current behaviour is fine. Yes, they are in some sense the same operation, but it's probably better for `ptrace` to always return a Qobj, rather than to suddenly return a complex number when a particular set of arguments is used.

File: qutip/tests/core/test_ptrace.py (hunk 1 above)
Comment (AGaliciaMartinez) [reply qutip/tests/core/test_ptrace.py:58]: I was not very clear what the problem was from the error message so it took me a little bit of digging to figure it out.

File: qutip/tests/core/data/test_ptrace.py (hunk 2 above)
Comment (AGaliciaMartinez) [reply qutip/tests/core/data/test_ptrace.py:92]: I guess so. I wont change it then.

File: qutip/core/data/ptrace.pyx [hunk 4]
```diff
@@ -17,17 +17,28 @@ __all__ = [
...
     cdef Py_ssize_t i
     dims = np.atleast_1d(dims).astype(idxint_dtype).ravel()
     sel = np.atleast_1d(sel).astype(idxint_dtype)
+    sel.sort()
+
     if sel.ndim != 1:
         raise ValueError("Selection must be one-dimensional")
-    sel.sort()
+
+    for d in dims:
```
[the remark this answers is not in the document]
Comment (AGaliciaMartinez) [reply qutip/core/data/ptrace.pyx:32]: I ended up adding a  `_check_inputs` function that performs all the checks and included the preparation step explicitly in each specialisation (they were only three lines after all). Hope it is more clear now.

**Comment by AGaliciaMartinez** [comment]
@hodgestar, thank you for the review! I believe all the suggestions were applied. Let me know if any other change is needed.

File: qutip/core/data/ptrace.pyx [hunk 5]
```diff
@@ -17,23 +17,28 @@ __all__ = [
...
+        raise ValueError(f"the input matrix shape, {shape} and the"
+                         f" dimension argument, {dims}, are not compatible.")
     if sel.ndim != 1:
         raise ValueError("Selection must be one-dimensional")
-    sel.sort()
+
+    for d in dims:
+        if d<1:
+            raise ValueError("dimensions must be greated than zero but where"
+                             f" dims={dims}.")
```
Comment (hodgestar) [inline qutip/core/data/ptrace.pyx:35]:
```suggestion
    if any(d < 1 for d in dims):
        raise ValueError("dimensions must be greated than zero but where"
                         f" dims={dims}.")
```

File: qutip/core/data/ptrace.pyx [hunk 6]
```diff
@@ -82,8 +87,12 @@ cdef inline void _i2_k_t(idxint N, idxint[:, ::1] tensor_table, idxint out[2]):
 
 
 cpdef CSR ptrace_csr(CSR matrix, object dims, object sel):
-    _check_shape(matrix)
-    dims, sel = _prepare_inputs(dims, sel)
+    dims = np.atleast_1d(dims).astype(idxint_dtype).ravel()
+    sel = np.atleast_1d(sel).astype(idxint_dtype)
+    sel.sort()
```
Comment (hodgestar) [inline qutip/core/data/ptrace.pyx:92]: Could we move these into `_check_inputs` somehow? My concern is that it is an error to call `_check_inputs` without running these lines (e.g. `dims` might be very nested) and it would be easy for us to forget this later.  It would require returning `dims` and `sel` and maybe changing the function name to `_parse_inputs` or something that suggests that inputs are parsed and modified.

**Review by hodgestar — CHANGES_REQUESTED** [review]
Two more small suggestions!

**Review by hodgestar — APPROVED** [review]
